controllers/machinedeletionremediation_controller.go

```diff
@@ -90,67 +90,45 @@ func (r *MachineDeletionRemediationReconciler) Reconcile(ctx context.Context, re
 		return ctrl.Result{}, err
 	}
 
-	annotations := remediation.GetAnnotations()
-	if annotations != nil {
-		machineNameNamespace, exists := annotations[MachineNameNamespaceAnnotation]
-		if exists {
-			machineName, machineNamespace, err := extractNameAndNamespace(machineNameNamespace, remediation.GetName())
-			if err != nil {
-				log.Error(err, "could not get Machine data from remediation", "remediation", remediation.GetName(), "annotation", machineNameNamespace)
-				return ctrl.Result{}, nil
-			}
+	log.Info("Machine Deletion Remediation CR found", "name", remediation.GetName())
 
-			ok, err := r.verifyMachineIsDeleted(ctx, machineName, machineNamespace, remediation.Name)
-			if err != nil {
-				return ctrl.Result{}, err
-			}
-			if !ok {
-				return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
-			}
+	var machine *unstructured.Unstructured
+	if machine, err = r.getMachine(remediation); err != nil {
+		if apiErrors.IsNotFound(err) {
 			return ctrl.Result{}, nil
 		}
-	}
-
-	var node *v1.Node
-	if node, err = r.getNodeFromMdr(remediation); err != nil {
+		log.Error(err, "Unexpected error fetching Machine from Node", "node", remediation.Name)
 		return ctrl.Result{}, err
 	}
 
-	log.Info("MDR CR and affected Node found", "node", node.GetName())
+	log.Info("node-associated machine found", "node", remediation.Name, "machine", machine.GetName())
 
-	var machine *unstructured.Unstructured
-	if machine, err = r.buildMachineFromNode(node); err != nil {
-		r.Log.Error(err, noMachineFoundError, "node name", node.Name)
-		return ctrl.Result{}, err
-	}
-
-	if !hasControllerOwner(machine) {
-		log.Info("ignoring remediation of node-associated machine: the machine has no controller owner", "machine", machine.GetName(), "node name", remediation.Name)
-		return ctrl.Result{}, nil
+	if !machine.GetDeletionTimestamp().IsZero() {
+		// Machine deletion already requested.
+		// Log deletion progress until the Machine exists
+		log.Info(postponedMachineDeletionInfo, "node", remediation.Name, "machine", machine.GetName(), "machine status.phase", getMachineStatusPhase(machine))
+		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
 	}
 
-	log.Info("node-associated machine found", "node", node.Name, "machine", machine.GetName())
+	if hasControllerOwner(machine) {
+		// save Machine's name and namespace to follow its deletion phase
+		if err = r.saveMachineNameNs(ctx, remediation, machine.GetName(), machine.GetNamespace()); err != nil {
```

Comment:
we can save an update call when we check if the annotation already exists first, either here or in saveMachineNameNs. Can happen when this call succeeds, but machine deletion below fails